source/kaleidic/mongo_gridfs.d

```diff
@@ -0,0 +1,62 @@
+module kaleidic.mongo_gridfs;
+
+import kaleidic.mongo_standalone;
+
+// IN:     mongo            - active connection to the MongoDB instance.
+//         gridfsBucketName - full path to the GridFS parent of collections .files and .chunks.
+//         id               - the file object ID, i.e., it's "_id" value.
+// RETURN: The entire file contents in an array.
+// THROWS: File not found, or on corrupt properties.
+
+ubyte[] gridfsReadFile(MongoConnection mongo, const string gridfsBucketName, const ObjectId id) {
+    import std.outbuffer;
+
+    // Get the file size.
+    OP_REPLY reply =
+        mongo.query(gridfsBucketName ~ ".files", 0, 1, document([bson_value("_id", id)]));
+    if (reply.errorCode != 0) {
+        throw new Exception(reply.errorMessage);
+    }
+    if (reply.documents.length != 1) {
+        throw new Exception("Requested file not found.");
+    }
+    long fileSize = reply.documents[0]["length"].get!long;
+    long chunkSize = reply.documents[0]["chunkSize"].get!long;
+
+    OutBuffer contents = new OutBuffer();
+    contents.reserve(fileSize);
+
+    // Loop for each chunk and append to our buffer.
+    int numChunks = cast(int)(fileSize / chunkSize + 1);
```

Comment:
this could be off-by-one when fileSize is a multiple of chunkSize.

You could fix this with integer calculations (`(fileSize-1)/chunkSize+1` and add a check before that fileSize > 0) or with floating point and rounding/ceiling, but that could introduce imprecision with huge numbers.

Alternatively wouldn't it also be possible to query for all the chunks without `n` and iterate over them? This would save MongoDB query calls as we already have a cursor then which we can iterate on multiple times. (though a single reply from the server might contain multiple chunks and be larger than we want it to be in memory, but I think it's possible to limit that and what it's doing automatically might not even be an issue)

Reply:
Whoops, nice find, how embarrassing.

I initially wrote code to query all the chunks at once and iterate through them, sending them to the OutBuffer.  But I wasn't absolutely sure that the results would be sorted on `n`.  They probably are, but I wasn't certain, and I don't see a way to enforce the sorting using this interface (though you can with other Mongo drivers)... @adamdruppe ?

And yeah, as you mention, doing a single query to fetch all the chunks would mean the entire file would be in memory twice, which would probably be OK 99% of the time, still could be an issue for very large files.  I have a feeling there might be better ways to do it, but I'm not sure this library has all the functionality we'd need and/or I can't really work it out yet, since me knowledge of Mongo internals is negligible (and TBH I'd rather keep it that way :wink:).

Reply:
actually queries in MongoDB don't put the whole file in (application) memory at once, you get fed the chunks while iterating (the op_reply in this library implements the cursor and performs the lazy fetching using ranges - so you can use it transparently in foreach)

The MongoDB docs say the file IDs are monotonously incrementing, but no mention about the chunk IDs. However you could add a MongoDB-side `sort` to your query.

I'm not sure whether MongoDB will load the full file into memory, that might be the case - what do other drivers do here actually?

Reply:
Before a few days ago I'd only ever heard of MongoDB as 'one of those nosql databases'.  I really don't know. :)

Reply:
ok found the spec here, there is other interesting stuff in there so sharing this while reading: https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst

Reply:
The spec says

> implementers retrieve all chunks with files_id equal to id, sorted in ascending order on “n”.

as well as

> As drivers stream the stored file they MUST check that each chunk received is the next expected chunk (i.e. it has the expected "n" value) and that the data field is of the expected length

and

> when downloading a zero length stored file the driver MUST NOT issue a query against the chunks collection

so I believe they intend us to use a single sorted query here.